stan/math/opencl/copy.hpp

```diff
@@ -52,8 +52,9 @@ inline matrix_cl<T> to_matrix_cl(const Eigen::Matrix<T, R, C>& src) {
      */
     cl::Event copy_event;
     const cl::CommandQueue& queue = opencl_context.queue();
-    queue.enqueueWriteBuffer(dst.buffer(), CL_FALSE, 0, sizeof(T) * dst.size(),
-                             src.data(), &dst.write_events(), &copy_event);
+    queue.enqueueWriteBuffer(dst.buffer(), opencl_context.in_order(), 0,
```

Comment:
While I did not test this method in that bug demo It needs same kind of changes as matrix_cl constructors. Addeing a simiar test for this could also be benefitial.

Reply:
I tried this with the following test: 

```
  MatrixXd a = MatrixXd::Random(N, N);
  MatrixXd b = MatrixXd::Random(N, N);
  matrix_cl<double> c_cl = to_matrix_cl(a + b);  // the problem
```
and that produces the following error: 
```
test/unit/math/opencl/copy_test.cpp:50:28: error: no matching function for call to 'to_matrix_cl'
  matrix_cl<double> c_cl = to_matrix_cl(a + b);  // the problem
                           ^~~~~~~~~~~~
./stan/math/opencl/copy.hpp:248:21: note: candidate template ignored: requirement
      'is_arithmetic<CwiseBinaryOp<scalar_sum_op<double, double>, const Matrix<double, -1, -1, 0, -1,
      -1>, const Matrix<double, -1, -1, 0, -1, -1> > >::value' was not satisfied [with T =
      Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<double, double>, const Eigen::Matrix<double,
      -1, -1, 0, -1, -1>, const Eigen::Matrix<double, -1, -1, 0, -1, -1> >]
inline matrix_cl<T> to_matrix_cl(const T& src) {
                    ^
./stan/math/opencl/copy.hpp:40:21: note: candidate template ignored: could not match 'Matrix' against
      'CwiseBinaryOp'
inline matrix_cl<T> to_matrix_cl(const Eigen::Matrix<T, R, C>& src) {
                    ^
1 error generated.
```
So the bug cant happen in this case. We will need to support this in the future but that is not part of this bugfix PR.

Reply:
I think you can probably work around this error by doing : ` matrix_cl<double> c_cl = to_matrix_cl((a + b).eval());`.

Reply:
Yep, that fails.